File: ServerPhp/App/Library/Package.php
<?php
namespace Wechat\App\Library;

use Wechat\App\Enums\OpCode;
use Workerman\Connection\TcpConnection;

/**
 * 数据包
 */
class Package
{
    /** @var null|TcpConnection 链接 */
    protected $connection = null;

    /** @var array 数据包内容 */
    protected $package = [
        'appId' => null,  // 客户端的应用ID，字符串，最长32位
        'appKey' => null,  // 客户端的应用KEY，字符串，最长32位
        'package' => [  // 数据包信息
            'uniqueId' => null,  // 数据包唯一ID，字符串，最长32位
            'timestamp' => null,  // 发送时间，毫秒级时间戳，字符串。13位
            'wechatId' => null,  // 接收指令的目标微信客户端ID，每开一个微信客户端带有一个唯一ID，无目标客户端则为null，字符串或null
            'opCode' => null,  // 操作指令，16进制数值
            'body' => null,  // 业务主体对象
        ],
    ];

    /**
     * Package constructor.
     */
    public function __construct()
    {

    }

    /**
     * 发送数据给客户端
     * @return bool|null
     * @throws \ErrorException
     */
    public function send()
    {
        // 检查必传参数
        $result = $this->check();
        if (!$result) {
            throw new \ErrorException('Package->send: 缺少必传参数！');
        }
        // 检查是否设置连接对象
        if (!$this->connection) {
            throw new \ErrorException('Package->send: 未设置连接对象！');
        }
        // 获取context
        $context = $this->getContext();
        return $this->connection->send($context);
    }

    /**
     * 设置客户端连接对象
     * @param TcpConnection $connection
     */
    public function setConnection(TcpConnection $connection)
    {
        $this->connection = $connection;
    }

    /**
     * 获取客户端连接对象
     * @return null|TcpConnection
     */
    public function getConnection(): TcpConnection
    {
        return $this->connection;
    }

    /**
     * 解包
     * @param $json
     */
    public function setConText($json)
    {
        try {
            $package = json_decode($json, true);

            $this->setAppId($package['appId'] ?? null);  // // 客户端的应用ID，字符串，最长32位
            $this->setAppKey($package['appKey'] ?? null);  // // 客户端的应用KEY，字符串，最长32位
            $this->setUniqueId($package['package']['uniqueId'] ?? null);  // 数据包唯一ID，字符串，最长32位
            $this->setTimestamp($package['package']['timestamp'] ?? null);  // 发送时间，毫秒级时间戳，数值
            $this->setWechatId($package['package']['wechatId'] ?? null);  // 接收指令的目标微信客户端ID，每开一个微信客户端带有一个唯一ID，无目标客户端则为null，字符串或null
            $this->setOpCode($package['package']['opCode'] ?? null);  // 操作指令，16进制数值
            $this->setBody($package['package']['body'] ?? null);  // 业务主体对象
        } catch (\ErrorException $exception) {
            Tools::log('Error： function setConText. ['.$exception->getCode().'] ' . $exception->getMessage());
        }
    }

    /**
     * 封包
     * @return string
     */
    public function getContext()
    {
        // 如果没设置时间戳，自动设置
        $this->createTimestamp();
        // 如果没设置消息ID，自动设置
        $this->createUniqueId();
        // 返回json数据
        return json_encode($this->package);
    }

    /**
     * 检查包的必传参数
     * @return bool
     */
    public function check()
    {
        // 检查应用ID
        $appId = $this->getAppId();
        if (is_null($appId)) {
            return false;
        }
        // 检查应用KEY
        $appKey = $this->getAppKey();
        if (is_null($appKey)) {
            return false;
        }
        // 检查操作指令
        $opCode = $this->getOpCode();
        if (is_null($opCode) || !in_array($opCode, OpCode::$allowOpCode)) {
            return false;
        }
        return true;
    }

    /**
     * 生成时间戳 - 毫秒级
     * @param bool $autoSet
     * @return string
     */
    public function createTimestamp($autoSet = true)
    {
        $timestamp = Tools::timestamp();
        if ($autoSet) {
            $this->setTimestamp($timestamp);
        }
        return (string)$timestamp;
    }

    /**
     * 生成包ID
     * @param bool $autoSet
     * @return string
     */
    public function createUniqueId($autoSet = true)
    {
        $uniqueId = strtoupper(md5(Tools::timestamp() . rand(100000, 999999)));
        if ($autoSet) {
            $this->setUniqueId($uniqueId);
        }
        return $uniqueId;
    }

    /**
     * @return string
     */
    public function getUniqueId()
    {
        return $this->package['package']['uniqueId'];
    }

    /**
     * @param string $uniqueId
     */
    public function setUniqueId($uniqueId = null)
    {
        $this->package['package']['uniqueId'] = $uniqueId;
    }

    /**
     * @return string
     */
    public function getTimestamp()
    {
        return (string)$this->package['package']['timestamp'];
    }

    /**
     * @param string $timestamp
     */
    public function setTimestamp($timestamp): void
    {
        $this->package['package']['timestamp'] = (string)$timestamp;
    }

    /**
     * @return array
     */
    public function getBody()
    {
        $body = $this->package['package']['body'];
        $body = empty($body) ? [] : $body;
        return $body;
    }

    /**
     * @param array $body
     */
    public function setBody($body)
    {
        $body = empty($body) ? (object)[] : $body;
        $this->package['package']['body'] = $body;
    }

    /**
     * @return int
     */
    public function getOpCode()
    {
        return $this->package['package']['opCode'];
    }

    /**
     * @param int $opCode
     */
    public function setOpCode($opCode)
    {
        $this->package['package']['opCode'] = $opCode;
    }

    /**
     * @return string
     */
    public function getWechatId()
    {
        return $this->package['package']['wechatId'];
    }

    /**
     * @param string $wechatId
     */
    public function setWechatId($wechatId)
    {
        $wechatId = $wechatId ? $wechatId : null;
        $this->package['package']['wechatId'] = $wechatId;
    }

    /**
     * @return null
     */
    public function getAppId()
    {
        return $this->package['appId'];
    }

    /**
     * @param string $appId
     */
    public function setAppId($appId)
    {
        $this->package['appId'] = $appId;
    }

    /**
     * @return string
     */
    public function getAppKey()
    {
        return $this->package['appKey'];
    }

    /**
     * @param string $appKey
     */
    public function setAppKey($appKey)
    {
        $this->package['appKey'] = $appKey;
    }
}

File: WechatRobot/WechatRobot/Package.cpp
#include "stdafx.h"
#include "Common.h"
#include "PJson.h"
#include "OpCode.h"
#include "Package.h"
#include <iostream>
#include <Windows.h>
#include <string>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <atlstr.h>

using namespace rapidjson;
using namespace std;

/** @var array 允许的操作指令 */
int allowOpCode[] = {
	/** @var int 响应指令 */
	0x0,  // ws客户端准备完毕
	0x1,  // 操作成功
	0x2,  // 操作失败
	/** @var int 业务指令 */
	0x10,  // 新开一个微信客户端
	0x11,  // 退出一个微信
	0x12,  // 获取登录状态
	0x20,  // 发送文本消息
	0x21,  // 发送图片消息
	0x22,  // 发送附件消息
	0x23,  // 发送名片消息
	0x24,  // 发送xml消息
	0x30,  // wxid加好友
	0x31,  // v1加好友
	0x32,  // 删除好友
	0x33,  // v1+v2同意好友请求
	0x34,  // 好友列表
	0x35,  // 修改备注
	0x40,  // 创建群聊
	0x41,  // 修改群名称
	0x42,  // 发送群公告
	0x43,  // 获取群成员列表
	0x44,  // 拉好友入群
	0x45,  // 删除群成员
	0x46,  // 艾特群成员
	0x47,  // 退出群聊
	0x50,  // 收款
	0x90,  // 返回登录二维码地址
	0x91,  // 返回当前登录的微信号信息
	0x92,  // 返回接收的微信消息
};

Package::Package()
{
	this->json.SetObject(); // 将当前的Document设置为一个object，也就是说整个Document是一个Object类型的dom元素
}
// 解包
void Package::SetConText(char *jsonStr)
{
	this->json.Parse(jsonStr);

	// 客户端的应用ID，字符串，最长32位
	if (this->json["appId"].IsNull()) {
		this->SetAppId(NULL);
	}
	else {
		if (!this->json["appId"].IsString()) return;
		Value& appId = this->json["appId"];
		this->SetAppId((char*)appId.GetString());
	}

	// 客户端的应用KEY，字符串，最长32位
	if (!this->json.HasMember("appKey")) return;
	if (this->json["appKey"].IsNull()) {
		this->SetAppKey(NULL);
	}
	else {
		if (!this->json["appKey"].IsString()) return;
		Value& appKey = this->json["appKey"];
		this->SetAppKey((char*)appKey.GetString());
	}

	if (!this->json.HasMember("package")) return;
	if (!this->json["package"].IsObject()) return;
	rapidjson::Value& packageObject = this->json["package"];

	// 数据包唯一ID，字符串，最长32位
	if (!packageObject.HasMember("uniqueId")) return;
	if (!packageObject["uniqueId"].IsString()) return;
	Value& uniqueId = packageObject["uniqueId"];
	this->SetUniqueId((char*)uniqueId.GetString());

	// 发送时间，毫秒级时间戳，字符串，13位
	if (!packageObject.HasMember("timestamp")) return;
	if (!packageObject["timestamp"].IsString()) return;
	Value& timestamp = packageObject["timestamp"];
	this->SetTimestamp((char*)timestamp.GetString());

	// 接收指令的目标微信客户端ID，每开一个微信客户端带有一个唯一ID，无目标客户端则为null，字符串或null
	if (!packageObject.HasMember("wechatId")) return;
	if (packageObject["wechatId"].IsNull()) {
		this->SetWechatId(NULL);
	}
	else {
		if (!packageObject["wechatId"].IsString()) return;
		Value& wechatId = packageObject["wechatId"];
		this->SetWechatId((char*)wechatId.GetString());
	}
	
	// 操作指令，16进制数值
	if (!packageObject.HasMember("opCode")) return;
	if (!packageObject["opCode"].IsInt()) return;
	Value& opCode = packageObject["opCode"];
	this->SetOpCode(opCode.GetInt());

	// 业务主体对象
	if (!packageObject.HasMember("body")) return;
	Value& body = packageObject["body"];
	if (packageObject["body"].IsNull()) {
		body.SetNull();
		this->SetBody(body);
	}
	else {
		if (!packageObject["body"].IsObject()) return;
		//rapidjson::Value& bodyObject = packageObject["body"];
		this->SetBody(packageObject["body"]);
	}
}
// 封包
rapidjson::StringBuffer Package::GetConText()
{
	// 客户端的应用ID，字符串，最长32位
	char *appId = this->GetAppId();
	rapidjson::Value appIdString(rapidjson::kStringType);
	if (!appId) {
		string sAppId = ReadConfig("app_id");
		char cAppId[33] = {0};
		sprintf_s(cAppId, sizeof(cAppId), "%s", sAppId.c_str());
		appIdString.SetString(cAppId, 32);
	}
	else {
		appIdString.SetString(appId, 32);
	}
	if (this->json.HasMember("appId")) {
		this->json.RemoveMember("appId");
	}
	this->json.AddMember("appId", appIdString, this->json.GetAllocator());

	// 客户端的应用KEY，字符串，最长32位
	char *appKey = this->GetAppKey();
	rapidjson::Value appKeyString(rapidjson::kStringType);
	if (!appKey) {
		string sAppKey = ReadConfig("app_key");
		char cAppKey[33] = { 0 };
		sprintf_s(cAppKey, sizeof(cAppKey), "%s", sAppKey.c_str());
		appKeyString.SetString(cAppKey, 32);
	}
	else {
		appKeyString.SetString(appKey, 32);
	}
	if (this->json.HasMember("appKey")) {
		this->json.RemoveMember("appKey");
	}
	this->json.AddMember("appKey", appKeyString, this->json.GetAllocator());

	// 初始化包数据
	rapidjson::Value packageObject(rapidjson::kObjectType);

	// 数据包唯一ID，字符串，最长32位
	char *uniqueId = this->GetUniqueId();
	rapidjson::Value uniqueIdString(rapidjson::kStringType);
	if (!uniqueId) {
		string uniqueId = CreateUniqueId();
		char cUniqueId[33] = { 0 };
		sprintf_s(cUniqueId, 33, "%s", uniqueId.c_str());
		uniqueIdString.SetString(cUniqueId, strlen(cUniqueId));
	}
	else {
		uniqueIdString.SetString(uniqueId, 32);
	}
	if (packageObject.HasMember("uniqueId")) {
		packageObject.RemoveMember("uniqueId");
	}
	packageObject.AddMember("uniqueId", uniqueIdString, this->json.GetAllocator());

	// 发送时间，毫秒级时间戳，数值
	char *timestamp = this->GetTimestamp();
	rapidjson::Value timestampString(rapidjson::kStringType);
	if (timestamp == NULL) {
		long long intTimestamp = CreateTimestamp();
		char sTimestamp[64] = { 0 };
		sprintf_s(sTimestamp, sizeof(sTimestamp), "%lld", intTimestamp);
		timestampString.SetString(sTimestamp, strlen(sTimestamp));
	}
	else {
		timestampString.SetString(timestamp, strlen(timestamp));
	}
	if (packageObject.HasMember("timestamp")) {
		packageObject.RemoveMember("timestamp");
	}
	packageObject.AddMember("timestamp", timestampString, this->json.GetAllocator());

	// 接收指令的目标微信客户端ID，每开一个微信客户端带有一个唯一ID，无目标客户端则为null，字符串或null
	char *wechatId = this->GetWechatId();
	if (wechatId == NULL) {
		rapidjson::Value wechatIdString(rapidjson::kNullType);
		wechatIdString.SetNull();
		if (packageObject.HasMember("wechatId")) {
			packageObject.RemoveMember("wechatId");
		}
		packageObject.AddMember("wechatId", wechatIdString, this->json.GetAllocator());
	}
	else {
		rapidjson::Value wechatIdString(rapidjson::kStringType);
		wechatIdString.SetString(wechatId, strlen(wechatId));
		if (packageObject.HasMember("wechatId")) {
			packageObject.RemoveMember("wechatId");
		}
		packageObject.AddMember("wechatId", wechatIdString, this->json.GetAllocator());
	}

	// 操作指令，16进制数值
	int opCode = this->GetOpCode();
	rapidjson::Value opCodeNumber(rapidjson::kNumberType);
	opCodeNumber.SetInt(opCode);
	if (packageObject.HasMember("opCode")) {
		packageObject.RemoveMember("opCode");
	}
	packageObject.AddMember("opCode", opCodeNumber, this->json.GetAllocator());

	// 业务主体对象
	rapidjson::Value *bodyObject = this->GetBody();
	if (packageObject.HasMember("body")) {
		packageObject.RemoveMember("body");
	}
	packageObject.AddMember("body", *bodyObject, this->json.GetAllocator());

	// 赋值包数据
	if (this->json.HasMember("package")) {
		this->json.RemoveMember("package");
	}
	this->json.AddMember("package", packageObject, this->json.GetAllocator());

	// 生成字符串
	rapidjson::StringBuffer buffer;
	Writer<rapidjson::StringBuffer> writer(buffer);
	this->json.Accept(writer);

	return buffer;
}
// 检查包的必传参数
bool Package::Check()
{
	// 检查应用ID
	char *appId = this->GetAppId();
	if (appId == NULL || strlen(appId) <= 0) {
		return FALSE;
	}

	// 检查应用KEY
	char *appKey = this->GetAppKey();
	if (appKey == NULL || strlen(appKey) <= 0) {
		return FALSE;
	}

	// 检查操作指令
	int opCode = this->GetOpCode();
	bool checkOpCodeResult = FALSE;
	for (int i = 0; i < sizeof(allowOpCode); i++) {
		if (opCode == allowOpCode[i]) {
			checkOpCodeResult = TRUE;
			break;
		}
	}
	if (checkOpCodeResult == FALSE) {
		return FALSE;
	}

	return TRUE;
}
/*==============================================================*/
// 获取客户端的应用ID
char *Package::GetAppId()
{
	return this->appId;
}
// 设置客户端的应用ID
void Package::SetAppId(char *appId)
{
	this->appId = appId;
}
// 获取客户端的应用KEY
char *Package::GetAppKey()
{
	return this->appKey;
}
// 设置客户端的应用KEY
void Package::SetAppKey(char *appKey)
{
	this->appKey = appKey;
}
/*==============================================================*/
// 获取数据包唯一ID
char *Package::GetUniqueId()
{
	return this->uniqueId;
}
// 设置数据包唯一
void Package::SetUniqueId(char *uniqueId)
{
	this->uniqueId = uniqueId;
}
/*==============================================================*/
// 获取发送时间
char *Package::GetTimestamp()
{
	return this->timestamp;
}
// 设置发送时间
void Package::SetTimestamp(char *timestamp)
{
	this->timestamp = timestamp;
}
/*==============================================================*/
// 获取接收指令的目标微信客户端ID
char *Package::GetWechatId()
{
	return this->wechatId;
}
// 设置接收指令的目标微信客户端ID
void Package::SetWechatId(char *wechatId)
{
	this->wechatId = wechatId;
}
/*==============================================================*/
// 获取操作指令
int Package::GetOpCode()
{
	return this->opCode;
}
// 设置操作指令
void Package::SetOpCode(int opCode)
{
	this->opCode = opCode;
}
/*==============================================================*/
// 获取业务主体对象
rapidjson::Value *Package::GetBody()
{
	return &body;
}
// 设置业务主体对象
void Package::SetBody(rapidjson::Value &body)
{
	this->body = body;
}